qsimcirq/qsim_circuit.py

```diff
@@ -212,45 +276,100 @@ def to_matrix(op: cirq.ops.GateOperation):
+
+    def to_matrix(op: cirq.ops.GateOperation):
+      mat = cirq.unitary(op.gate, None)
+      if mat is None:
+          return NotImplemented
+      
+      return cirq.ops.MatrixGate(mat).on(*op.qubits)
+
+    qubit_to_index_dict = {q: i for i, q in enumerate(ordered_qubits)}
+    time_offset = 0
+    for moment in self:
+      moment_length = 0
+      ops_by_gate = []
+      ops_by_mix = []
+      ops_by_channel = []
+      # Capture ops of each type in the appropriate list.
+      for qsim_op in moment:
+        if cirq.has_unitary(qsim_op) or cirq.is_measurement(qsim_op):
+          oplist = cirq.decompose(
+            qsim_op, fallback_decomposer=to_matrix, keep=has_qsim_kind)
+          ops_by_gate.append(oplist)
+          moment_length = max(moment_length, len(oplist))
+          pass
+        elif cirq.has_mixture(qsim_op):
+          ops_by_mix.append(qsim_op)
+          moment_length = max(moment_length, 1)
+          pass
+        elif cirq.has_channel(qsim_op):
+          ops_by_channel.append(qsim_op)
+          moment_length = max(moment_length, 1)
+          pass
+        else:
+          raise ValueError(f'Encountered unparseable op: {qsim_op}')
+
+      # Gates must be added in time order.
+      for gi in range(moment_length):
+        # Handle gate output.
+        for gate_ops in ops_by_gate:
+          if gi >= len(gate_ops):
+            continue
+          qsim_op = gate_ops[gi]
+          time = time_offset + gi
+          gate_kind = _cirq_gate_kind(qsim_op.gate)
+          add_op_to_circuit(qsim_op, time, qubit_to_index_dict, qsim_ncircuit)
+        # Handle mixture output.
+        for mixture in ops_by_mix:
+          mixdata = []
+          for prob, mat in cirq.mixture(mixture):
+            square_mat = np.reshape(mat, (int(np.sqrt(mat.size)), -1))
+            unitary = cirq.is_unitary(square_mat)
+            # Package matrix into a qsim-friendly format.
+            mat = np.reshape(mat, (-1,)).astype(np.complex64, copy=False)
+            mixdata.append((prob, mat.view(np.float32), unitary))
+          qubits = [qubit_to_index_dict[q] for q in mixture.qubits]
+          qsim.add_channel(time_offset, qubits, mixdata, qsim_ncircuit)
+        # Handle channel output.
+        for channel in ops_by_channel:
+          chdata = []
+          for i, mat in enumerate(cirq.channel(channel)):
+            square_mat = np.reshape(mat, (int(np.sqrt(mat.size)), -1))
+            unitary = cirq.is_unitary(square_mat)
+            singular_vals = np.linalg.svd(square_mat)[1]
+            lower_bound_prob = min(singular_vals) ** 2
+            # Package matrix into a qsim-friendly format.
+            mat = np.reshape(mat, (-1,)).astype(np.complex64, copy=False)
+            chdata.append((lower_bound_prob, mat.view(np.float32), unitary))
+          qubits = [qubit_to_index_dict[q] for q in channel.qubits]
+          qsim.add_channel(time_offset, qubits, chdata, qsim_ncircuit)
+      time_offset += moment_length
```

Comment:
It seems like doing this math in python and then passing these matrices through the translation might incur a lot of uneseccesary overhead vs just doing this SVD on the qsim side. Maybe this would be a good time to add Eigen as a dependency ?

Reply:
I won't say this is infeasible, but it would add a degree of complexity - both to this PR and to the repository as a whole, since we would need to add dependency on Eigen (likely similar to the existing googletest submodule).

How much overhead do you see this adding? I wasn't able to find a concrete answer on Eigen vs. numpy performance in a quick search, but with most operators being 1- or 2-qubit I wouldn't expect there to be much difference.

Reply:
On a related note: during design we considered putting SVD calculation in C++ so that our C++ users could access it, but IIRC we decided against it under the assumption that C++ users would prefer precalculating their lower bounds for improved performance.

Reply:
I think this shouldn't be a significant overhead, especially, when this is done once for a large number of trajectories.